### PR TITLE
[VC02-4325]-Timeslots sanity check

### PR DESCRIPTION
**Pull request by EwaGrela**

### CODE REVIEW

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 1]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+def test_30_days_single_timeslots(api_url, api_key, product_id, endpoint, adapter_func, version=1):
+    '''Checking timeslots'''
+    start = datetime.utcnow().date()
+    end = start + timedelta(days=30)
+
+    raw_response, response = client(f'{api_url}/v{version}/products/{product_id}/{endpoint}', api_key, {
+        'start': start.isoformat(),
+        'end': end.isoformat(),
+    })
+
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:55]: First parse the results and turn them into a list of `Timeslot` objects:
```py
timeslots = adapter_func(raw_response, response)
```
then use them as objects.

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 2]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+    dates_from_response = [result["date"] for result in response]
+    unique_dates_from_response = set(dates_from_response)
+    if len(unique_dates_from_response) == len(dates_from_response):
+        starts_ends = [{"start": item["start"], "end": item["end"]} for item in response]
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:59]: 💋 
```py
unique_starts_ends = {f'{t.start}-{t.end}' for t in timeslots}
if len(unique_starts_ends) == 1:
    raise FailedTest(
          message='If a product contains only a single timeslot at the same time every day, then please implement it as a non-timesloted product',
          response=raw_response,
      )
```

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 3]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+def test_30_days_timeslots_duplicates(api_url, api_key, product_id, endpoint, adapter_func, version=1):
+    '''Checking timeslots'''
+    start = datetime.utcnow().date()
+    end = start + timedelta(days=30)
+
+    raw_response, response = client(f'{api_url}/v{version}/products/{product_id}/{endpoint}', api_key, {
+        'start': start.isoformat(),
+        'end': end.isoformat(),
+    })
+
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:78]:
```py
timeslots = adapter_func(raw_response, response)
```

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 4]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+    response_copy = deepcopy(response)
+    for item in response_copy:
+        item.pop("variants")
+
+    unique_timeslots_no_variants = [i for n, i in enumerate(response_copy) if i not in response_copy[n + 1:]]
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:83]: This isn't clear to me what's happening here. Also usually `i` is used for the index number with `enumerate`.

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 5]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+
+    if response_copy != unique_timeslots_no_variants:
+        raise FailedTest(
+        message='Timeslots with/without variants cannot have identical content',
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:87]: wrong indent

File: supplier_api_tester/supplier_api_tester/utils/availability.py [hunk 6]
```diff
@@ -43,6 +43,59 @@ def test_next_30_days(api_url, api_key, product_id, endpoint, adapter_func, vers
...
+        response=raw_response,
+        )
+    unique_whole_timeslots =  [i for n, i in enumerate(response) if i not in response[n + 1:]]
+
+    if response != unique_whole_timeslots:
+        raise FailedTest(
+        message='Timeslots cannot be identical',
```
Comment (trojkat) [inline supplier_api_tester/supplier_api_tester/utils/availability.py:94]: wrong indent

File: supplier_api_tester/supplier_api_tester/utils/availability.py (hunk 4 above)
Comment (EwaGrela) [reply supplier_api_tester/supplier_api_tester/utils/availability.py:83]: What I want to do is perform two checks.
First if there are no 'similar' timeslots, meaning - they have the same date, start, end but variants differ - these are not allowed.
I am creating a copy of response containing timeslots, then I am removing variants from the copy.
Then I am making a list with unique records and then I am comparing copy and unique records.
If they are not identical - there must be duplicates.

**Review by trojkat — DISMISSED** [review]

**Review by trojkat — APPROVED** [review]

**Review by Loreq12 — APPROVED** [review]

**Pull request merged**